soc/xtensa/intel_adsp/ace_v1x/power.c

```diff
@@ -26,29 +26,194 @@
+{
+	__asm__ __volatile__("wsr.excsave3 %0"
+		: : "a"(excsave3) : "memory");
+}
+
+#define BITMASK_FROM_RANGE_UINT32(from, to) ((0xFFFFFFFF << (from)) & \
+					    (0xFFFFFFFF >> (sizeof(uint32_t) * 8 - 1 - (to))))
+#define LPSCTL_BATTR_MASK                   BITMASK_FROM_RANGE_UINT32(12, 16)
+#define LPSRAM_MAGIC_VALUE 0x13579BDF
+
+struct lpsram_header {
+	uint32_t alt_reset_vector;
+	uint32_t adsp_lpsram_magic;
+	void *lp_restore_vector;
+	uint32_t reserved;
+	uint32_t slave_core_vector;
+	uint8_t rom_bypass_vectors_reserved[0xC00 - 0x14];
+};
+
+enum PowerFlow {
+	D0_TO_D0i3,
+	D0i3_TO_D0
+};
+
+struct core_state {
+	enum PowerFlow flow;
+	uint32_t intenable;
+	uint32_t thread_ptr;
+	uint32_t vecbase;
+	uint32_t excsave2;
+	uint32_t excsave3;
+};
+
+static struct core_state core_desc[CONFIG_MP_NUM_CPUS] = { 0 };
+
+void dsp_pg_restore(void);
+
+__asm__("dsp_pg_restore:\n\t"
+	"  movi  a0, 0\n\t"
+	"  movi  a1, 1\n\t"
+	"  wsr   a1, WINDOWSTART\n\t"
+	"  wsr   a0, WINDOWBASE\n\t"
+	"  rsync\n\t"
+	"  movi a3, 0x20\n\t"
+	"  rsr a4, intenable\n\t"
+	"  or a4, a4, a3\n\t"
+	"  wsr a4, intenable\n\t"
+	"  rsync\n\t"
+	"  wsr a3, interrupt\n\t"
+	"  rsync\n\t"
+	"  loop1:\n\t"
+	"    nop\n\t"
+	"    j loop1\n\t");
+
+#define SOFTWARE_INTERRUPT_NUMBER 5
+
+void software_irq_handler(void *arg)
+{
+	uint32_t core_id = processor_id_reg();
+
+	if (core_desc[core_id].flow == D0_TO_D0i3) {
+		struct lpsram_header *lpsheader =
+			(struct lpsram_header *) DT_REG_ADDR(DT_NODELABEL(sram1));
+		lpsheader->adsp_lpsram_magic = LPSRAM_MAGIC_VALUE;
+		lpsheader->lp_restore_vector = (uint32_t) &dsp_pg_restore;
+		core_desc[core_id].flow = D0i3_TO_D0;
+		z_xtensa_cache_flush_inv_all();
+		k_cpu_idle();
+	} else if (core_desc[core_id].flow == D0i3_TO_D0) {
+		write_threadptr(core_desc[core_id].thread_ptr);
+		write_vecbase(core_desc[core_id].vecbase);
+		write_excsave2(core_desc[core_id].excsave2);
+		write_excsave3(core_desc[core_id].excsave3);
+		__asm__ volatile("rsync");
+	}
+}
+
+static void power_gate(uint32_t core_id)
+{
+	core_desc[core_id].flow = D0_TO_D0i3;
+	DFDSPBRCP.bootctl[core_id].bctl &= ~DFDSPBRCP_BCTL_WAITIPPG;
+	DFDSPBRCP.bootctl[core_id].bctl &= ~DFDSPBRCP_BCTL_WAITIPCG;
+	ACE_PWRCTL->wpdsphpxpg &= ~BIT(core_id);
+	if (core_id == 0) {
+		uint32_t battr = DFDSPBRCP.bootctl[core_id].battr & (~LPSCTL_BATTR_MASK);
+
+		battr |= (DFDSPBRCP_BATTR_LPSCTL_RESTORE_BOOT & LPSCTL_BATTR_MASK);
+		DFDSPBRCP.bootctl[core_id].battr = battr;
+	}
+
+	core_desc[core_id].intenable = read_intenable();
+	core_desc[core_id].thread_ptr = read_threadptr();
+	core_desc[core_id].vecbase = read_vecbase();
+	core_desc[core_id].excsave2 = read_excsave2();
+	core_desc[core_id].excsave3 = read_excsave3();
+	soc_cpus_active[core_id] = false;
+	z_xt_ints_off(0xffffffff);
+	IRQ_CONNECT(SOFTWARE_INTERRUPT_NUMBER, 0, software_irq_handler, 0, 0);
+	z_xtensa_irq_enable(SOFTWARE_INTERRUPT_NUMBER);
+	z_xt_set_intset(1<<SOFTWARE_INTERRUPT_NUMBER);
```

Comment:
Why this is here ? That should be done once and not every time the system goes idle ...